tools/cpp/modules_tools/common/common.cc

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.h"
+#include "json.hpp"
+
+#include <iostream>
+
+ModuleDep parse_ddi(std::istream &ddi_stream) {
+  ModuleDep dep{};
+  std::string ddi_string((std::istreambuf_iterator<char>(ddi_stream)),
+                         std::istreambuf_iterator<char>());
+  JsonValue data = parse_json(ddi_string);
+  if (!data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+
+  auto rules_data = data.as_object().at("rules");
+  if (!rules_data.is_array()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto rules = rules_data.as_array();
+  if (rules.size() > 1) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  if (rules.empty()) {
+    return dep;
+  }
+  auto rule_data = rules[0];
+  if (!rule_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto rule = rule_data.as_object();
+  auto provides_data = rule["provides"];
+  if (!provides_data.is_array()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  auto provides = provides_data.as_array();
+  if (provides.size() > 1) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  if (provides.size() == 1) {
+    auto provide_data = provides[0];
+    if (!provide_data.is_object()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    auto name_data = provide_data.as_object().at("logical-name");
+    if (!name_data.is_string()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    dep.gen_bmi = true;
+    dep.name = name_data.as_string();
+  }
+  auto requires_data = rule["requires"];
+  if (!requires_data.is_object()) {
+    std::cerr << "bad ddi" << std::endl;
+    std::exit(1);
+  }
+  for (const auto &item_data : requires_data.as_array()) {
+
+    if (!item_data.is_object()) {
+      std::cerr << "bad ddi" << std::endl;
+      std::exit(1);
+    }
+    auto name_data = item_data.as_object().at("logical-name");
```

Comment:
You might want to reject anything that says it is a header unit here as unsupported.

Reply:
I tested scanning C++ source files using header units with clang-scan-deps and GCC 14, and they both failed the scan. Header unit support in both Clang and GCC is still quite lacking. Therefore, I believe we can disregard the logic related to header units for now. Since the scan fails, there won't be any valid DDI file content.

tested with `demo.cc`
```
import <iostream>;
int main() { std::cout << "Hello World.\n"; }
```

- clang-scan-deps
```
CC=/opt/homebrew/Cellar/llvm/18.1.5/bin/clang
scanner=/opt/homebrew/Cellar/llvm/18.1.5/bin/clang-scan-deps

$scanner -format=p1689 -- $CC -std=c++20 -c demo.cc
```
got error
```
error: header file <iostream> (aka '/opt/homebrew/Cellar/llvm/18.1.5/bin/../include/c++/v1/iostream') cannot be imported because it is not known to be a header unit
{
  "revision": 0,
  "rules": [],
  "version": 1
}
```

- gcc-14
```
gcc-14 -E -x c++ -fmodules-ts -fdeps-file=out.tmp -fdeps-format=p1689r5 demo2.cc \
-MT demo.cc.o.ddi -MD -MF demo.cc.o.ddi.d -fdeps-file=demo.cc.o.ddi
```
got error

```
# 0 "demo.cc"
# 0 "<built-in>"
# 0 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 0 "<command-line>" 2
# 1 "demo2.cc"
import  "/usr/include/c++/14/iostream";
In module imported at demo.cc:1:1:
/usr/include/c++/14/iostream: error: failed to read compiled module: No such file or directory
/usr/include/c++/14/iostream: note: compiled module file is 'gcm.cache/./usr/include/c++/14/iostream.gcm'
/usr/include/c++/14/iostream: note: imports must be built before being imported
/usr/include/c++/14/iostream: fatal error: returning to the gate for a mechanical issue
compilation terminated.
```